src/Nethermind/Nethermind.Monitoring/Metrics/MetricsController.cs

```diff
@@ -18,43 +20,61 @@ public partial class MetricsController : IMetricsController
         private readonly Dictionary<Type, (string DictName, IDictionary<string, long> Dict)> _dynamicPropCache = new();
         private readonly HashSet<Type> _metricTypes = new();
 
         public readonly Dictionary<string, Gauge> _gauges = new();
+        private readonly bool _useCounters;
 
         public void RegisterMetrics(Type type)
         {
-            EnsurePropertiesCached(type);
-            foreach ((PropertyInfo propertyInfo, string gaugeName) in _propertiesCache[type])
+            if (_metricTypes.Add(type) == false)
             {
-                _gauges[gaugeName] = CreateMemberInfoMectricsGauge(propertyInfo);
+                return;
             }
 
-            foreach ((FieldInfo fieldInfo, string gaugeName) in _fieldsCache[type])
+            Meter meter = new(type.Namespace);
+
+            EnsurePropertiesCached(type);
+            foreach ((MemberInfo member, string gaugeName, Func<double> observer) in _membersCache[type])
             {
-                _gauges[gaugeName] = CreateMemberInfoMectricsGauge(fieldInfo);
-            }
+                if (_useCounters)
+                {
+                    CreateDiagnosticsMetricsObservableGauge(meter, member, observer);
+                }
 
-            _metricTypes.Add(type);
+                _gauges[gaugeName] = CreateMemberInfoMetricsGauge(member);
+            }
         }
 
-        private Gauge CreateMemberInfoMectricsGauge(MemberInfo propertyInfo)
+        private static Gauge CreateMemberInfoMetricsGauge(MemberInfo member)
         {
-            var staticLabels = propertyInfo
+            Dictionary<string, string> staticLabels = member
                 .GetCustomAttributes<MetricsStaticDescriptionTagAttribute>()
                 .ToDictionary(
                     attribute => attribute.Label,
                     attribute => GetStaticMemberInfo(attribute.Informer, attribute.Label));
 
-            string description = propertyInfo.GetCustomAttribute<DescriptionAttribute>()?.Description;
-            string name = BuildGaugeName(propertyInfo);
+            string description = member.GetCustomAttribute<DescriptionAttribute>()?.Description;
+            string name = BuildGaugeName(member);
 
             return CreateGauge(name, description, staticLabels);
         }
 
+        private static ObservableInstrument<double> CreateDiagnosticsMetricsObservableGauge(Meter meter, MemberInfo member, Func<double> observer)
+        {
+            string description = member.GetCustomAttribute<DescriptionAttribute>()?.Description;
+            string name = member.GetCustomAttribute<DataMemberAttribute>()?.Name ?? member.Name;
+
+            if (member.GetCustomAttribute<CounterMetricAttribute>() != null)
+            {
+                return meter.CreateObservableCounter(name, observer, description: description);
+            }
+
+            return meter.CreateObservableGauge(name, observer, description: description);
```

Comment:
shouldn't this explicitly be based on the attribute?

Reply:
I don't think so. `Gauge` is the most blunt `Instrument` as it reports just the value as is. The intent for attribute was to make it for future folks adding metrics so that they pause and select either `++/--` semantic of the counter of `just a value` of a gauge. I think we should report everything as `Gauge` by default. I hope it clarifies it.